documentation/optee_design.md

```diff
@@ -523,8 +523,117 @@ implementation, as explained in [crypto.md].
+
+Static TAs do not benefit from the 'GP Core Internal API' support
+specified by the GP TEE specs; Static TAs can only use the OP-TEE OS internal
+APIs and routines.
+
+Pseudo-TAs have the same privileged view of Secure World as the OP-TEE OS code
+itself.  For complex TAs, that may not be desirable.
+
+In most cases a real, dynamically loaded TA is the best choice instead of adding
+your code directly to the OP-TEE OS.  However if you decide your application
+is best handled directly in OP-TEE OS like this, you can look at
+`core/arch/arm/sta/stats.c` as a template and just add your static TA based on
+that to the `sub.mk` in the same directory.
+
+### Trusted Applications
+
+Trusted Applications (TAs) are applications dynamically loaded by OP-TEE
+OS in the Secure World when something in the REE wants to talk to that
+particular application UUID.  It's similar to the way the Linux
+kernel can dynamically load kernel modules, although unlike with Linux, in
+OP-TEE TAs actually run at a lower CPU privilege level than OP-TEE OS code.
+
+Because the TAs are signed by the same key that built the OP-TEE OS, they are
+able to be stored in the untrusted REE filesystem, and tee-supplicant will
+take care of passing them to be checked and loaded by the Secure World OP-TEE
+OS.  Again this is simular to Linux kernel module signature checking.
+
+Trusted Application benefit from the `GP Core Internal API` as specified by
+the 'GP Specs'.
+
+Trusted Application consist of a cleartext signed ELF file, named from the UUID
+of the TA and the suffix ".ta".
+
+They are built separately from the OP-TEE OS boot-time blob, although when
+they are built they use the same build system, and are signed with the key
+from the build of the original OP-TEE OS blob.
+
+## Special treatment of Trusted Applications
+
+### Syscalls
+
+Dynamically loaded TAs are not directly bound to function exports in the OP-TEE
+OS blob, both because the TA code is kept at arm's length by executing at a
+different privilege level, and because TAs direct binding to addresses in the
+OS would require upgrades of all TAs synchronusly with upgrades of the OP-TEE
+OS blob.  Instead, the resolution of OP-TEE OS exports in the TA is done at
+runtime.
+
+OP-TEE does this by using syscalls, the same kind of way as the Linux kernel
+provides a stable API for its userland programs.  TAs are written to use
+syscall wrappers to access functions exported from OP-TEE OS, so this all
+happens automatically when a TA wants to use an api exported from
+OP-TEE OS.
+
+Static / pseudo TAs and anything else directly built into OP-TEE OS do not
+require going through a syscall interface, since they can just link directly
+as they are directly part of the OS.
+
+Syscalls are provided already for all public exports from OP-TEE OS that a
+Dynamic TA is expected to use, so you only need to take care about this if
+you will add new exported from OP-TEE OS that TAs will want to use.
```

Comment:
sure, i understand. and thank you for all the contributions.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`